zcash_client_sqlite/src/wallet.rs

```diff
@@ -519,28 +521,201 @@ pub(crate) fn get_balance(
-    let balance = conn.query_row(
-        "SELECT SUM(value) FROM sapling_received_notes
-        INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
-        WHERE account = ? AND spent IS NULL AND transactions.block <= ?",
-        [u32::from(account), u32::from(anchor_height)],
-        |row| row.get(0).or(Ok(0)),
+    min_confirmations: u32,
+) -> Result<Option<WalletSummary>, SqliteClientError> {
+    let chain_tip_height = match scan_queue_extrema(conn)? {
+        Some((_, max)) => max,
+        None => {
+            return Ok(None);
+        }
+    };
+
+    let birthday =
+        wallet_birthday(conn)?.expect("If a scan range exists, we know the wallet birthday.");
+
+    let fully_scanned_height =
+        block_fully_scanned(conn)?.map_or(birthday - 1, |m| m.block_height());
+    let summary_height = chain_tip_height + 1 - min_confirmations;
+
+    let sapling_scan_progress = {
+        // Compute the number of notes scanned directly from the blocks table
+        let fully_scanned = conn
+            .query_row(
+                "SELECT SUM(sapling_output_count)
+                 FROM blocks
+                 WHERE height <= :fully_scanned_height",
+                named_params![":fully_scanned_height": u32::from(fully_scanned_height)],
+                |row| row.get::<_, Option<u64>>(0),
+            )
+            .optional()?
+            .flatten()
+            .unwrap_or(0);
+
+        if fully_scanned_height == chain_tip_height && fully_scanned > 0 {
+            Ok(Some(Ratio::new(fully_scanned, fully_scanned)))
+        } else {
+            // We don't have complete information on how many outputs will exist in the shard at
+            // the chain tip without having scanned the chain tip block, so we overestimate by
+            // computing the maximum possible of notes directly from the shard indices.
+            //
+            // TODO: it would be nice to be able to reliably have the size of the commitment tree
+            // at the chain tip without having to have scanned that block.
+            conn.query_row(
+                "SELECT MIN(shard_index), MAX(shard_index)
+                 FROM sapling_tree_shards
+                 WHERE subtree_end_height > :birthday_height
+                 OR subtree_end_height IS NULL",
+                named_params![":birthday_height": u32::from(birthday)],
+                |row| {
+                    let min_idx = row.get::<_, Option<u64>>(0)?;
+                    let max_idx = row.get::<_, Option<u64>>(1)?;
+                    Ok(min_idx.zip(max_idx).map(|(min, max)| {
+                        let max_tree_size = (max + 1) << SAPLING_SHARD_HEIGHT;
+                        let min_tree_size = min << SAPLING_SHARD_HEIGHT;
+                        Ratio::new(fully_scanned, max_tree_size - min_tree_size)
+                    }))
+                },
+            )
+            .optional()
+            .map(|opt| opt.flatten())
+        }
```

Comment:
This whole thing is a bit of a nightmare, because we don't have any source of note commitment tree sizes at the wallet birthday or at the chain tip. If we could reliably get those pieces of information, then this could be much better but this is a reasonable minimum-viable implementation.